crates/storage/provider/src/providers/blockchain_provider.rs

```diff
@@ -1473,25 +1473,31 @@ mod tests {
     use std::sync::Arc;
 
     use reth_chain_state::{ExecutedBlock, NewCanonicalChain};
-    use reth_primitives::B256;
-    use reth_storage_api::{BlockHashReader, BlockNumReader, HeaderProvider};
+    use reth_db::{test_utils::TempDatabase, DatabaseEnv};
+    use reth_primitives::{BlockNumberOrTag, SealedBlock, B256};
+    use reth_storage_api::{BlockHashReader, BlockNumReader, BlockReaderIdExt, HeaderProvider};
     use reth_testing_utils::generators::{self, random_block_range};
 
-    use crate::{providers::BlockchainProvider2, test_utils::create_test_provider_factory};
+    use crate::{
+        providers::BlockchainProvider2, test_utils::create_test_provider_factory, CanonChainTracker,
+    };
 
-    #[test]
-    fn test_block_hash_reader() -> eyre::Result<()> {
+    type ArcedTempDatabase = Arc<TempDatabase<DatabaseEnv>>;
+
+    const BLOCK_COUNT: usize = 10;
+
+    fn provider_with_random_blocks(
+    ) -> eyre::Result<(BlockchainProvider2<ArcedTempDatabase>, Vec<SealedBlock>)> {
         let mut rng = generators::rng();
+        let limit = BLOCK_COUNT as u64 - 1;
+        let blocks = random_block_range(&mut rng, 0..=limit, B256::ZERO, 0..1);
 
         let factory = create_test_provider_factory();
-
-        // Generate 10 random blocks
-        let blocks = random_block_range(&mut rng, 0..=10, B256::ZERO, 0..1);
+        let provider_rw = factory.provider_rw()?;
 
         let mut blocks_iter = blocks.clone().into_iter();
 
         // Insert first 5 blocks into the database
-        let provider_rw = factory.provider_rw()?;
         for block in (0..5).map_while(|_| blocks_iter.next()) {
```

Comment:
this doesn't respect the `BLOCK_COUNT` constant. Let's accept two arguments instead: `database_blocks: usize, in_memory_blocks: usize`, and return `database_blocks: Vec<SealedBlock>, in_memory_blocks: Vec<SealedBlock>`